packages/datetime/src/dateRangeInput.tsx

```diff
@@ -46,22 +69,82 @@ export class DateRangeInput extends AbstractComponent<IDateRangeInputProps, {}>
         },
     };
 
+    public constructor(props: IDateRangeInputProps, context?: any) {
+        super(props, context);
+        this.state = {
+            isOpen: false,
+            selectedEnd: moment(null),
+            selectedStart: moment(null),
+        };
+    }
+
     public render() {
+        const popoverContent = (
+            <DateRangePicker
+                onChange={this.handleDateRangePickerChange}
+                maxDate={this.props.maxDate}
+                minDate={this.props.minDate}
+                value={this.getSelectedRange()}
+            />
+        );
+
         // allow custom props for each input group, but pass them in an order
         // that guarantees only some props are overridable.
         return (
-            <div className={Classes.CONTROL_GROUP}>
-                <InputGroup
-                    placeholder="Start date"
-                    {...this.props.startInputProps}
-                    inputRef={this.refHandlers.startInputRef}
-                />
-                <InputGroup
-                    placeholder="End date"
-                    {...this.props.endInputProps}
-                    inputRef={this.refHandlers.endInputRef}
-                />
-            </div>
+            <Popover
+                autoFocus={false}
+                content={popoverContent}
+                enforceFocus={false}
+                inline={true}
+                isOpen={this.state.isOpen}
+                onClose={this.handlePopoverClose}
+                position={Position.TOP_LEFT}
+            >
+                <div className={Classes.CONTROL_GROUP}>
+                    <InputGroup
+                        placeholder="Start date"
+                        {...this.props.startInputProps}
+                        inputRef={this.refHandlers.startInputRef}
+                        onClick={this.handleInputClick}
+                        onFocus={this.handleInputFocus}
+                    />
+                    <InputGroup
+                        placeholder="End date"
+                        {...this.props.endInputProps}
+                        inputRef={this.refHandlers.endInputRef}
+                        onClick={this.handleInputClick}
+                        onFocus={this.handleInputFocus}
+                    />
+                </div>
+            </Popover>
         );
     }
+
+    private handleDateRangePickerChange = (selectedRange: DateRange) => {
+        const [selectedStart, selectedEnd] = selectedRange.map(fromDateToMoment);
+        this.setState({ selectedStart, selectedEnd });
+        Utils.safeInvoke(this.props.onChange, selectedRange);
```

Comment:
There was a discussion here on what we pass into callbacks when there are no dates selected: https://github.com/palantir/blueprint/pull/524

You don't have the controlled case yet, so it's not yet relevant but there will be the same issue with keeping the callback value sane between `null / undefined / moment(null)` especially when it's tempting for consumers to use `this.setState({ value })` directly